tower-balance/src/p2c/service.rs

```diff
@@ -21,26 +23,47 @@ use tower_service::Service;
 /// [finagle]: https://twitter.github.io/finagle/guide/Clients.html#power-of-two-choices-p2c-least-loaded
 /// [p2c]: http://www.eecs.harvard.edu/~michaelm/postscripts/handbook2001.pdf
 #[derive(Debug)]
-pub struct Balance<D: Discover> {
+pub struct Balance<D: Discover, Req> {
     discover: D,
 
-    endpoints: IndexMap<D::Key, D::Service>,
+    ready: IndexMap<D::Key, D::Service>,
+
+    unready: stream::FuturesUnordered<CancelReady<D::Key, D::Service, Req>>,
+    cancelations: IndexMap<D::Key, oneshot::Sender<()>>,
 
     /// Holds an index into `endpoints`, indicating the service that has been
     /// chosen to dispatch the next request.
-    ready_index: Option<usize>,
+    next_ready_index: Option<usize>,
 
     rng: SmallRng,
 }
 
-impl<D: Discover> Balance<D> {
+#[derive(Debug)]
+struct CancelReady<K, S, Req> {
```

Comment:
This name confused me. If I am reading this correctly, this struct waits on a service to become ready, and is cancellable. The name made me think the action taken by this struct was to cancel readiness.

Reply:
Related to this, I'd like to see some docs on `CancelReady` to explain why it's needed and how it's used + maintained.

Reply:
Maybe `RemovableService` is a better name?

Reply:
Landed on `UnreadyService` to be consistent with the (updated) field name `unready_services`.